src/upgrade/check.rs

```diff
@@ -3,24 +3,60 @@ use ferinth::structures::version::{Version, VersionFile};
 use furse::structures::file_structs::File;
 use octocrab::models::repos::{Asset, Release};
 
+fn is_jar_file(asset_name: &str) -> bool {
+    asset_name.contains("jar")
+}
+
+fn is_not_source(asset_name: &str) -> bool {
+    !asset_name.contains("source")
+}
+
+fn game_version_check_contain(game_version: Option<&str>, asset_name: &str) -> bool {
+    game_version
+        .map(|game_version| asset_name.contains(game_version))
+        // select latest asset if version check is disabled
+        .unwrap_or(true)
+}
+
+fn game_version_check_exact(game_version: Option<&str>, version: &str) -> bool {
+    game_version
+        .map(|game_version| version == game_version)
+        // select latest asset if version check is disabled
+        .unwrap_or(true)
+}
+
+fn mod_loader_check_contain(mod_loader: Option<&ModLoader>, asset_name: &str) -> bool {
+    mod_loader
+        .map(|mod_loader| {
+            asset_name
+                .split('-')
+                .any(|loader| loader == mod_loader.to_string().as_str())
+        })
+        // select latest asset if mod loader check is disabled
+        .unwrap_or(true)
+}
+
+fn mod_loader_check_exact(mod_loader_to_check: Option<&ModLoader>, mod_loader: &str) -> bool {
+    mod_loader_to_check
+        .map(|loader| loader.to_string().as_str() == mod_loader)
+        // select latest asset if mod loader check is disabled
+        .unwrap_or(true)
+}
+
 /// Get the latest compatible file from `files`
 pub fn curseforge<'a>(
     files: &'a mut [File],
     game_version_to_check: Option<&str>,
     mod_loader_to_check: Option<&ModLoader>,
 ) -> Option<&'a File> {
-    // Make the newest files come first
-    files.sort_unstable_by_key(|file| file.file_date);
-    files.reverse();
+    // Sort files in non increasing order
+    files.sort_unstable_by(|file1, file2| file2.file_date.cmp(&file1.file_date));
 
     // Immediately select the newest file if check is disabled, i.e. *_to_check is None
     files.iter().find(|file| {
-        file.game_versions.iter().any(|version| {
-            game_version_to_check.is_none() || version == game_version_to_check.unwrap()
-        }) && file.game_versions.iter().any(|mod_loader| {
-            mod_loader_to_check.is_none()
-                || Ok(mod_loader_to_check.unwrap())
-                    == ModLoader::try_from(mod_loader.as_ref()).as_ref()
+        file.game_versions.iter().any(|asset| {
+            game_version_check_exact(game_version_to_check, asset)
+                && mod_loader_check_exact(mod_loader_to_check, asset)
```

Comment:
I think I've found out why this doesn't work.

CF groups the mod loader and game version in the same list, but you're checking _both_ the game version and mod loader for the same element. You have to perform both checks separately, because even though they're in the same vector they are different things.

Reply:
Oh I see, I miss interpreted something here, Now I understood. Thanks

Reply:
```patch
@@ -54,10 +54,13 @@ pub fn curseforge<'a>(
 
     // Immediately select the newest file if check is disabled, i.e. *_to_check is None
     files.iter().find(|file| {
-        file.game_versions.iter().any(|asset| {
-            game_version_check_exact(game_version_to_check, asset)
-                && mod_loader_check_exact(mod_loader_to_check, asset)
-        })
+        file.game_versions
+            .iter()
+            .any(|asset| game_version_check_exact(game_version_to_check, dbg!(asset)))
+            && file
+                .game_versions
+                .iter()
+                .any(|asset| mod_loader_check_exact(mod_loader_to_check, dbg!(asset)))
     })
 }
```